### ISSUE TITLE
Ramses tform utility

### ISSUE THREAD

**Issue by Martin-Rey** [pynbody/pynbody#520]
Hi,

I have come across two bugs while using the Ramses utility get_tform.

1) If the path to the snapshot has a "/" at the end (following an autocomplete with tab for example), the utility would fail due to its parsing mechanics here:
https://github.com/pynbody/pynbody/blob/cb18c8d0ba847975046db550ec281d3c3cd48e25/pynbody/analysis/ramses_util.py#L437-L444

in which the "/" is taken out in the parent_dir variable, but the top.filename variable is used to extract the 12 characters mapping to the output name. You can see an example following, for which the error follows from this try/catch block:
https://github.com/pynbody/pynbody/blob/cb18c8d0ba847975046db550ec281d3c3cd48e25/pynbody/analysis/ramses_util.py#L455-L464

```
In [1]: s = pynbody.load("./Halo1459_fiducial_fblim/output_00032/")
/Users/mrey/anaconda3/lib/python3.6/site-packages/pynbody-0.47-py3.6-macosx-10.7-x86_64.egg/pynbody/snapshot/ramses.py:731: UserWarning: Using field 'tform' to distinguish stars. If this is wrong, try editing your config.ini, section [ramses], entry particle-distinguisher.
  warnings.warn("Using field %s to distinguish stars. If this is wrong, try editing your config.ini, section [ramses], entry particle-distinguisher."%pb_name)

In [2]: pynbody.analysis.ramses_util.get_t(s)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-3ac456de44d7> in <module>()
----> 1 pynbody.analysis.ramses_util.get_t(s)

AttributeError: module 'pynbody.analysis.ramses_util' has no attribute 'get_t'

In [3]: pynbody.analysis.ramses_util.get_tform(s)
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
~/anaconda3/lib/python3.6/site-packages/pynbody-0.47-py3.6-macosx-10.7-x86_64.egg/pynbody/analysis/ramses_util.py in get_tform(sim, part2birth_path, cpu_range)
    456             birth_file = FortranFile('%s/output_%s/birth_%s.out%05d' %
--> 457                                      (parent_dir, top._timestep_id, top._timestep_id, i + 1))
    458         except IOError:

~/anaconda3/lib/python3.6/site-packages/scipy/io/_fortran.py in __init__(self, filename, mode, header_dtype)
     91         else:
---> 92             self._fp = open(filename, '%sb' % mode)
     93 

FileNotFoundError: [Errno 2] No such file or directory: './Halo1459_fiducial_fblim/o/output_00032/birth_00032.out00001'

During handling of the above exception, another exception occurred:

FileNotFoundError                         Traceback (most recent call last)
<ipython-input-3-c877739c8e47> in <module>()
----> 1 pynbody.analysis.ramses_util.get_tform(s)

~/anaconda3/lib/python3.6/site-packages/pynbody-0.47-py3.6-macosx-10.7-x86_64.egg/pynbody/analysis/ramses_util.py in get_tform(sim, part2birth_path, cpu_range)
    462             with open(os.devnull, 'w') as fnull:
    463                 subprocess.call([part2birth_path, '-inp', 'output_%s' % top._timestep_id],
--> 464                                 stdout=fnull, stderr=fnull)
    465             # part2birth put the files in output_<top._timestep_id>
    466             birth_file = FortranFile('%s/output_%s/birth_%s.out%05d' %

~/anaconda3/lib/python3.6/subprocess.py in call(timeout, *popenargs, **kwargs)
    265     retcode = call(["ls", "-l"])
    266     """
--> 267     with Popen(*popenargs, **kwargs) as p:
    268         try:
    269             return p.wait(timeout=timeout)

~/anaconda3/lib/python3.6/subprocess.py in __init__(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, encoding, errors)
    707                                 c2pread, c2pwrite,
    708                                 errread, errwrite,
--> 709                                 restore_signals, start_new_session)
    710         except:
    711             # Cleanup if the child failed starting.

~/anaconda3/lib/python3.6/subprocess.py in _execute_child(self, args, executable, preexec_fn, close_fds, pass_fds, cwd, env, startupinfo, creationflags, shell, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite, restore_signals, start_new_session)
   1342                         if errno_num == errno.ENOENT:
   1343                             err_msg += ': ' + repr(err_filename)
-> 1344                     raise child_exception_type(errno_num, err_msg, err_filename)
   1345                 raise child_exception_type(err_msg)
   1346 

FileNotFoundError: [Errno 2] No such file or directory: '$HOME/ramses/trunk/ramses/utils/f90/part2birth': '$HOME/ramses/trunk/ramses/utils/f90/part2birth'

In [4]: s = pynbody.load("./Halo1459_fiducial_fblim/output_00032")
/Users/mrey/anaconda3/lib/python3.6/site-packages/pynbody-0.47-py3.6-macosx-10.7-x86_64.egg/pynbody/snapshot/ramses.py:731: UserWarning: Using field 'tform' to distinguish stars. If this is wrong, try editing your config.ini, section [ramses], entry particle-distinguisher.
  warnings.warn("Using field %s to distinguish stars. If this is wrong, try editing your config.ini, section [ramses], entry particle-distinguisher."%pb_name)

In [5]: pynbody.analysis.ramses_util.get_tform(s)
Out[5]: 
SimArray([0.95871874, 0.46736587, 0.46622571, ..., 0.66457786, 0.69522304,
          0.67728758], 'Gyr')
```

2) The tform utility makes use of the _info['ncpu'] snapshot attribute, which is a Ramses specific protected key:
https://github.com/pynbody/pynbody/blob/cb18c8d0ba847975046db550ec281d3c3cd48e25/pynbody/analysis/ramses_util.py#L432

When combining pynbody and Tangos in server mode, the `top` is a `RemoteSubSnap` class of Tangos  which is agnostic about the specific `pynbody.snapshot.SimSnap` type used. Calculations in server mode would then fail with the following error:
```
[  1] 2019-05-03 10:10:41,513 :    StarFormHistogram 0.1s | 0.0%
[  1] 2019-05-03 10:10:41,513 :             StarForm 0.0s | 0.0%
[  2] 2019-05-03 10:10:41,585 : Uncaught exception during property calculation <edge_tangos_properties.star_formation.StarFormHistogram object at 0x1a1d097e48> applied to <Halo 'Halo1459_fiducial_fblim/output_00032/halo_2' | NDM=64218 Nstar=0 Ngas=0>
Traceback (most recent call last):
  File "/Users/mrey/Documents/tangos/tangos/tools/property_writer.py", line 376, in _get_property_value
    result = property_calculator.calculate(snapshot_data, db_data)
  File "/Users/mrey/Documents/tangos/tangos/properties/pynbody/centring.py", line 66, in new_fn
    return fn(self, halo, existing_properties)
  File "/Users/mrey/Documents/edge-tangos-properties/edge_tangos_properties/star_formation.py", line 38, in calculate
    self._ensure_ramses_tform_are_corrected(simulation_data)
  File "/Users/mrey/Documents/edge-tangos-properties/edge_tangos_properties/star_formation.py", line 26, in _ensure_ramses_tform_are_corrected
    ru.get_tform(particle_data, part2birth_path=None, cpu_range=None)
  File "/Users/mrey/anaconda3/lib/python3.6/site-packages/pynbody-0.47-py3.6-macosx-10.7-x86_64.egg/pynbody/analysis/ramses_util.py", line 432, in get_tform
    ncpu = top._info['ncpu']
  File "/Users/mrey/anaconda3/lib/python3.6/site-packages/pynbody-0.47-py3.6-macosx-10.7-x86_64.egg/pynbody/snapshot/__init__.py", line 445, in __getattr__
    type(self).__name__, name))
AttributeError: 'RemoteSubSnap' object has no attribute '_info'
```
 
Note that calculations in default parallel mode succeed.

Overall, 1) can be fixed very easily but 2) suggest that these utilities have ramses specific knowledge, belonging inside the Ramses loader rather than as post-processing of the general `pynbody.snapshot.SimSnap` type.

Furthermore, utilities such as 
https://github.com/pynbody/pynbody/blob/cb18c8d0ba847975046db550ec281d3c3cd48e25/pynbody/analysis/ramses_util.py#L81
https://github.com/pynbody/pynbody/blob/cb18c8d0ba847975046db550ec281d3c3cd48e25/pynbody/analysis/ramses_util.py#L119
https://github.com/pynbody/pynbody/blob/cb18c8d0ba847975046db550ec281d3c3cd48e25/pynbody/analysis/ramses_util.py#L142

are now useless following the addition of the HOP catalogue as a core pynbody functionality and should be deleted.

Let me know what you think about the best way to clean/solve this.

Martin

**Comment by apontzen** [comment]
Thanks...

1) The path mishandling should hopefully be easy to patch -- please do so if it's causing you problems
2) Fundamentally this error reflects that the current implementation, where users need to call a particular utility routine, is pretty bizarre. I think it shouldn't be too much work to:
 * rename the `tform` field in ramses to `tform_raw` (or similar) 
 * create a derived array `tform` for `RamsesSnap` that calls the required utility. (Note that derived arrays are correctly called for remote snapshots in tangos)
3) I agree with deleting these utilities that are no longer required

**Comment by Martin-Rey** [comment]
Agreed for 1-2-3. 

I'll submit a PR next week that addresses these issues.

**Comment by apontzen** [comment]
Actually I suddenly realised my suggestion for (2) may not be such a great one -- as a _derived_ array would be called on the worker process, not the server process.

It probably needs to be implemented instead as a branch from `_load_array` inside `RamsesSnap`

**Comment by Martin-Rey** [comment]
Ah ok. By branch, do you mean a new special case in the large tree of if/elif/else like here:
https://github.com/pynbody/pynbody/blob/cb18c8d0ba847975046db550ec281d3c3cd48e25/pynbody/snapshot/ramses.py#L948-L974

**Comment by apontzen** [comment]
Yes, not pretty but I think it's probably necessary

**Comment by Martin-Rey** [comment]
Fixed by #521
